src/Draco.Compiler/Internal/Solver/ConstraintSolver_Rules.cs

```diff
@@ -1,388 +1,409 @@
+                    // Variadic part
+                    foreach (var (paramType, arg) in variadicPairs) AssignParameterToArgument(store, paramType, arg);
+                }
+                else
+                {
+                    foreach (var (param, arg) in chosen.Parameters.Zip(overload.Candidates.Arguments))
+                    {
+                        AssignParameterToArgument(store, param.Type, arg);
+                    }
+                }
+                // NOTE: This used to be an assignment, but again, I don't think that's in the scope of this constraint
+                // In all cases, return type is simple
+                UnifyAsserted(overload.ReturnType, chosen.ReturnType);
+                // Resolve promise
+                overload.CompletionSource.SetResult(chosen);
+            })
+            .Named("overload"),
+
+        // As a last resort, we try to drive forward the solver by trying to merge assignable constraints with the same target
+        // This is a common situation for things like this:
+        //
+        // var x = Derived();
+        // x = Base();
+        //
+        // In this case we try to search for the common type of Derived and Base, then assign that
+        Simplification(typeof(Assignable), typeof(Assignable))
+            .Guard((Assignable a1, Assignable a2) =>
+                SymbolEqualityComparer.AllowTypeVariables.Equals(a1.TargetType, a2.TargetType))
+            .Body((ConstraintStore store, Assignable a1, Assignable a2) =>
+            {
+                var targetType = a1.TargetType;
+                var commonType = this.AllocateTypeVariable();
+                store.Add(new CommonAncestor(
+                    locator: ConstraintLocator.Constraint(a2),
+                    commonType: commonType,
+                    alternativeTypes: [a1.AssignedType, a2.AssignedType]));
+                store.Add(new Assignable(
+                    locator: ConstraintLocator.Constraint(a2),
+                    targetType: targetType,
+                    assignedType: commonType));
+            })
+            .Named("merge_assignables"),
+
+        // As a last-last effort, we assume that a singular assignment means exact matching types
+        Simplification(typeof(Assignable))
+            .Body((ConstraintStore store, Assignable assignable) =>
+            {
+                // TODO: Is asserted correct here?
+                // Maybe just for type-variables?
+                UnifyAsserted(assignable.TargetType, assignable.AssignedType);
+            })
+            .Named("sole_assignable"),
+
+        // As a last-effort, if we see a common ancestor constraint with a single non-type-var, we
+        // assume that the common type is the non-type-var
+        // We also substitute all the type-vars with the common type
+        Simplification(typeof(CommonAncestor))
```

Comment:
What happen if we do: 
```typescript
var x = superType;
x.SuperMethod();
x = baseType;
```

Reply:
I assume you mean super to be deriver.
x will be inferred to be baseType, and `x.SuperMethod()` will be an error. I believe this is sensible behavior since whatever you choose, it would be an error.

Reply:
Yes that's the outcome we want but I wanted to see what the compiler would do here.